service/service.go

```diff
@@ -170,16 +169,75 @@ func NewService(sig *signal.Signal, cfg *Config, datadir string) (*Service, erro
 	return s, nil
 }
 
+func (s *Service) loop(ctx context.Context) {
+	executingRounds, _ := errgroup.WithContext(ctx)
+	defer executingRounds.Wait()
+
+	timer := time.NewTimer(0)
+	<-timer.C
+	defer timer.Stop()
+	for {
+		start := s.genesis.
+			Add(s.cfg.EpochDuration * time.Duration(s.openRound.Epoch())).
+			Add(s.cfg.PhaseShift)
+		if d := time.Until(start); d > 0 {
+			log.Info("Round %v waiting for execution to start for %v", s.openRoundID(), d)
+			timer.Reset(d)
+			select {
+			case <-timer.C:
+			case <-ctx.Done():
+				log.Info("service shutting down")
+				s.openRoundMutex.Lock()
+				s.openRound = nil
+				s.openRoundMutex.Unlock()
+				return
+			}
+		}
+
+		s.openRoundMutex.Lock()
+		if s.openRound.isEmpty() && !s.cfg.ExecuteEmpty {
+			s.openRoundMutex.Unlock()
+			continue
+		}
+
+		prevRound := s.openRound
+		open := time.Now()
+		s.newRound(ctx, prevRound.Epoch()+1)
+		s.openRoundMutex.Unlock()
+		log.Info("Round %v opened. took %v", s.openRoundID(), time.Since(open))
+
+		executingRounds.Go(func() error {
+			round := prevRound
+			if err := s.executeRound(ctx, round); err != nil {
+				s.asyncError(fmt.Errorf("round %v execution error: %v", round.ID, err))
+				return nil
```

Comment:
I assume this should return an error instead of `nil`?

Reply:
No, I specifically don't want a failed round to cancel the context and close the Service. An error is logged and Poet Service continues.